wpimath/src/main/java/edu/wpi/first/math/geometry/AllianceFlipper.java

```diff
@@ -0,0 +1,223 @@
+    ROTATIONALLY_MIRRORED {
+      public double flipX(double x) {
+        return activeYear.fieldLength - x;
+      }
+
+      public double flipY(double y) {
+        return activeYear.fieldWidth - y;
+      }
+
+      public double flipHeading(double heading) {
+        return Math.PI - heading;
+      }
+    };
+
+    /**
+     * Flips the X coordinate.
+     *
+     * @param x The X coordinate to flip.
+     * @return The flipped X coordinate.
+     */
+    public abstract double flipX(double x);
+
+    /**
+     * Flips the Y coordinate.
+     *
+     * @param y The Y coordinate to flip.
+     * @return The flipped Y coordinate.
+     */
+    public abstract double flipY(double y);
+
+    /**
+     * Flips the heading.
+     *
+     * @param heading The heading to flip.
+     * @return The flipped heading.
+     */
+    public abstract double flipHeading(double heading);
+  }
+
+  /**
+   * An interface for objects that can be flipped based on the current alliance.
+   */
+  public static interface Flippable<Self extends Flippable<Self>> {
+    /**
+     * Flips the object based on the supplied flipper.
+     */
+    public Self flip(Flipper flipper);
+
+    /**
+     * Flips the object based on the active flipper.
+     */
+    public default Self flip() {
+      return flip(getFlipper());
+    }
+
+    /**
+     * Flips the object if on the red alliance, otherwise returns the object unchanged.
+     */
+    @SuppressWarnings("unchecked")
+    public default Self flipIfRed() {
+        return onRed() ? flip() : (Self) this;
+    }
+  }
+
+  private static record YearInfo(Flipper flipper, double fieldLength, double fieldWidth) {}
+
+  private static final HashMap<Integer, YearInfo> flipperMap =
+      new HashMap<Integer, YearInfo>() {
+        {
+          put(2022, new YearInfo(Flipper.ROTATIONALLY_MIRRORED, 16.4592, 8.2296));
+          put(2023, new YearInfo(Flipper.VERTICALLY_MIRRORED, 16.54175, 8.0137));
+          put(2024, new YearInfo(Flipper.VERTICALLY_MIRRORED, 16.54175, 8.211));
+        }
+      };
+
+  private static YearInfo activeYear = flipperMap.get(2024);
+  private static BooleanSupplier onRed = () -> false;
+
+  /**
+   * Get the flipper that is currently active for flipping coordinates. It's reccomended not to
+   * store this locally as the flipper may change.
+   *
+   * @return The active flipper.
+   */
+  public static Flipper getFlipper() {
+    return activeYear.flipper;
+  }
+
+  /**
+   * Returns if you are on red alliance, if the alliance is unknown it will return false.
+   *
+   * @return If you are on red alliance.
+   */
+  public static boolean onRed() {
+    return onRed.getAsBoolean();
+  }
+
+  /**
+   * Returns if you are on blue alliance, if the alliance is unknown it will return true.
+   *
+   * @return If you are on blue alliance.
+   */
+  public static boolean onBlue() {
+    return !onRed.getAsBoolean();
```

Comment:
I don't think this should be part of the library. It doesn't buy the user / vendordep much over having the user control it.